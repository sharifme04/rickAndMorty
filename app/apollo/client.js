import {ApolloClient, InMemoryCache} from '@apollo/client';
import fetch from 'cross-fetch';
import {createHttpLink} from 'apollo-link-http';

// Initialize Apollo Client
const client = new ApolloClient({
  link: createHttpLink({
    uri: 'https://rickandmortyapi.com/graphql/',
    fetch: fetch,
  }),
  cache: new InMemoryCache(),
});
export default client;
