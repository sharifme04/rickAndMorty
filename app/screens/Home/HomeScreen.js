import React, {useState, useEffect} from 'react';
import {useQuery} from '@apollo/client';
import {
  FlatList,
  Text,
  TextInput,
  ActivityIndicator,
  SafeAreaView,
} from 'react-native';
import styles from './Style';
import {theme} from '../../theme/theme';
import RenderItem from './RenderItem';
import {GET_CHARACTERS} from '../../apollo/queries';

const HomeScreen = ({navigation}) => {
  const [search, setSearch] = useState('');
  const [page, setPage] = useState(1);
  const {loading, error, data} = useQuery(GET_CHARACTERS, {
    variables: {search, page},
  });

  const [total, setTotal] = useState(data?.characters?.results);

  useEffect(() => {
    if (page === 1) {
      return setTotal(data?.characters?.results);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data?.characters, page]);

  const handleLoadMore = () => {
    setPage(page + 1);
    setTotal([...total, ...data?.characters?.results]);
  };

  if (error) {
    return <Text>{error.message}</Text>;
  }
  return (
    <SafeAreaView style={styles.container}>
      <TextInput
        style={styles.search}
        onChangeText={(text) => setSearch(text)}
        placeholder="Please search here"
        value={search}
      />
      {loading ? (
        <ActivityIndicator size="large" color={theme.colors.headerBackground} />
      ) : (
        <FlatList
          data={total}
          renderItem={(item) => (
            <RenderItem {...item} navigation={navigation} />
          )}
          keyExtractor={(item, i) => (item.id + i).toString()}
          onEndReached={handleLoadMore}
          onEndReachedThreshold={0.5}
        />
      )}
    </SafeAreaView>
  );
};

export default HomeScreen;
