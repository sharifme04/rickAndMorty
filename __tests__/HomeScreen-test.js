import 'react-native';
import React from 'react';
import {MockedProvider} from '@apollo/client/testing';
import HomeScreen from '../app/screens/Home/HomeScreen';
import renderer from 'react-test-renderer';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

const fakeresult = [
  {
    episode: [],
    gender: 'Male',
    id: '1',
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    name: 'Rick Sanchez',
    species: 'Human',
    __typename: 'Character',
  },
  {
    episode: [
      {
        air_date: 'September 10, 2017',
        name: 'The Ricklantis Mixup',
        __typename: 'Episode',
      },
    ],
    gender: 'Male',
    id: '8',
    image: 'https://rickandmortyapi.com/api/character/avatar/8.jpeg',
    name: 'Adjudicator Rick',
    species: 'Human',
    __typename: 'Character',
  },
  {
    episode: [
      {
        air_date: 'April 7, 2014',
        name: 'Close Rick-counters of the Rick Kind',
        __typename: 'Episode',
      },
    ],
    gender: 'Male',
    id: '15',
    image: 'https://rickandmortyapi.com/api/character/avatar/15.jpeg',
    name: 'Alien Rick',
    species: 'Alien',
    __typename: 'Character',
  },
];

test('HomeScreen', () => {
  const tree = renderer
    .create(
      <MockedProvider addTypename={false}>
        <HomeScreen />
      </MockedProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('HomeScreen with result', () => {
  const tree = renderer
    .create(
      <MockedProvider addTypename={false}>
        <HomeScreen total={fakeresult} />
      </MockedProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('HomeScreen empty result ', () => {
  const tree = renderer
    .create(
      <MockedProvider addTypename={false}>
        <HomeScreen total={[]} />
      </MockedProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
