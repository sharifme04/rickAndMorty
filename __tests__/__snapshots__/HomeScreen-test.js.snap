// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Please search here"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "gray",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 40,
        "margin": 20,
        "textAlign": "center",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <ActivityIndicator
    animating={true}
    color="#f4511e"
    hidesWhenStopped={true}
    size="large"
  />
</RCTSafeAreaView>
`;

exports[`HomeScreen empty result  1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Please search here"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "gray",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 40,
        "margin": 20,
        "textAlign": "center",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <ActivityIndicator
    animating={true}
    color="#f4511e"
    hidesWhenStopped={true}
    size="large"
  />
</RCTSafeAreaView>
`;

exports[`HomeScreen with result 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Please search here"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "gray",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 40,
        "margin": 20,
        "textAlign": "center",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <ActivityIndicator
    animating={true}
    color="#f4511e"
    hidesWhenStopped={true}
    size="large"
  />
</RCTSafeAreaView>
`;
